Task1

LARCH1K opened the pull request:
# Проект: Telegram-bot для ведения личных финансов

## Команда: Петрушина Надежда, Овчинников Никита

### Задача №1
Реализовать скелет телеграмм-бота. Бот должен уметь:
* Идентифицировать пользователя (по tg id)
* Обеспечить возможность работы нескольких пользователей одновременно
* Показать справку по работе бота (/help)
* Записывать расходы ( - <Сумма> <Наименование> )
* Записывать доходы ( + <Сумма> <Наименование> )
* Показывать баланс (/balance)
* Показать список расходов (/expenses)
* Показать список доходов (/income)

В данной задаче не будет интеграции с Гугл-таблицами,
вместо этого будет реализация БД в оперативной памяти.

LARCH1K committed:
Добавление нового модуля и зависимостей

LARCH1K committed:
Реализация части функционала бота

* Идентификация пользователя:
- Создана структура данных для хранения необходимой информации в оперативной памяти.

* Реализация команд:
- Добавлен обработчик команды /expenses, который возвращает список всех расходов пользователя.
- Добавлен обработчик команды /income, который возвращает список всех доходов пользователя.

* Тесты
- Написаны тесты

hopenz committed:
Merge pull request #5 from hopenz/task1-ovchinnikov

Task1-ovchinnikov

hopenz committed:
Добавлены обработчики команд help и balance

* Добавила обработчик команды /balance, который возвращает текущий баланс пользователя.

* Реализовала обработчик команды /help, который отправляет пользователю список всех доступных команд и их описание.

hopenz committed:
Реализация обработчиков команд и сообщений

* Реализация обработчиков команд:
- Добавлен обработчик команды /balance, который возвращает текущий баланс пользователя;
- Добавлен обработчик команды /help, который отправляет пользователю список всех доступных команд и их описание.

* Реализация обработчиков сообщений:
- Добавлен обработчик для сообщений формата - <Сумма> <Наименование> для записи расходов;
- Добавлен обработчик для сообщений формата + <Сумма> <Наименование> для записи доходов.

* Тесты:
- Написаны тесты

LARCH1K committed:
Merge pull request #8 from hopenz/task1-petrushina

Task1 petrushina

YaEtoTui commented on bot/src/main/java/ru/naumen/bot/configuration/ApplicationConfig.java:
```diff
@@ -0,0 +1,19 @@
+@ConfigurationProperties(prefix = "app", ignoreUnknownFields = false)
+public record ApplicationConfig(
+        String telegramToken
+) {
```
у нас не С#. Это не ошибка а ремарка, форматируем вот так -> "String telegramToken) {" . Точнее будет правильнее

LARCH1K replied:
Учтём и исправим

vpyzhyanov replied:
Ребята, замечания нужны, не для красоты и не "для вида", а чтобы их **исправлять**

hopenz replied:
Принято! В дальнейшем будем править сразу)
Эти ремарки уже были исправлены на момент вчерашней проверки

YaEtoTui commented on bot/src/main/java/ru/naumen/bot/data/dao/inMemory/InMemoryBalanceDao.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Override
+    public Double getBalance(long chatId) {
+        return usersBalance.get(chatId);
```
Хм. возможно вы предусмотрели, но а если баланса нет в Map, то get() вызовется с ошибкой. Тогда уж лучше getOrDefault. он отдаст либо из Map, либо default value. Почему я его вспомнил, а потому что у вас Double, а не double. То есть вы подразумеваете такое в теории

YaEtoTui commented on bot/src/main/java/ru/naumen/bot/data/dao/inMemory/InMemoryBalanceDao.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Override
+    public Double getBalance(long chatId) {
+        return usersBalance.get(chatId);
```
лучше мне кажется getOrDefault тут, так как тут ладно если null. а если вы вызываете в BalanceService, то там возвращает из Double -> double. Мне кажется в будущем можете получить баг, так что лучше getOrDefault(), а там вставьте 0 как дефолт, как бы решите будущую проблему, на мой взгляд

LARCH1K replied:
Согласен, в дальнейшем исправим

vpyzhyanov replied:
Ребята, замечания нужны, не для красоты и не "для вида", а чтобы их **исправлять**

vpyzhyanov replied:
с замечанием полностью согласен, чем меньше используется null, тем меньше в программе неожиданных NPE

hopenz replied:
Вячеслав Юрьевич, так как Егор написал в телеграме, что это не ошибки, а просто ремарки, мы приняли к сведению  и со второй задачей хотели поправить

YaEtoTui replied:
Да, это ремарка, но лучше конечно поправить. От данной команды я услышал, что ремарка будет поправлено. 
В дальнейшем я буду писать тут, чтобы Вячеслав понимал.

vpyzhyanov replied:
> со второй задачей хотели поправить

Зачем ждать? Если не сделаете сейчас, потом 99,9% забудете об этом. Есть такая народная мудрость "когда-то потом = никогда"

hopenz replied:
Принято! В дальнейшем будем править сразу)
Эти ремарки уже были исправлены на момент вчерашней проверки

vpyzhyanov commented on pom.xml:
```diff
@@ -8,10 +8,52 @@
 
+    <packaging>pom</packaging>
+    <modules>
+        <module>bot</module>
```
Начнём со структуры проекта. Многомодульный проект состоящий из одного модуля - это вообще зачем? Не надо усложнять структуру на ровном месте. Если нет необходимости, то не надо усложнять. У вас, скорее всего, и не будет потребности в других модулях.

LARCH1K replied:
Проект переделан в одномодульный

vpyzhyanov commented on bot/.gitignore:
зачем вам несколько gitignore? В этом никакого смысла нет. Вообще-то репозиторий всегда **один**, даже в многомодульном проекте.

LARCH1K replied:
Убран вместе с дочернем модулем

vpyzhyanov commented on bot/.gitignore:
```diff
@@ -0,0 +1,33 @@
+HELP.md
```
Это осознанное игнорирование? Подозреваю, что нет. Значит надо убрать

LARCH1K replied:
Весь файл убран

vpyzhyanov commented on bot/.gitignore:
```diff
@@ -0,0 +1,33 @@
+.sts4-cache
+
+### IntelliJ IDEA ###
+.idea
```
![image](https://github.com/user-attachments/assets/8582acfa-fe13-4540-9100-c21deb0b4ab1)

(скрин из главного gitignore) Надо понимать, что игнорируя всю папку, вы игнорируете всё её содержимое

LARCH1K replied:
Лишние строки убраны

hopenz committed:
Fix mistakes

- Исправлено формартирование в ApplicationConfig
- В InMemoryBalanceDao в методе getBalance get заменён getOrDefault
- В BalanceService метод getBalance теперь возвращает Double вместо double

LARCH1K committed:
Merge pull request #10 from hopenz/task1-petrushina

Fix mistakes

hopenz commented:
Замечания исправлены)

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/controller/TgBot.java:
```diff
@@ -0,0 +1,75 @@
+ * обрабатывает входящие обновления от Telegram, используя цепочку фильтров.
+ */
+@Component
+public class TgBot extends TelegramBot {
```
Добро пожаловать на тёмную сторону. Своровали ответственность другого класса. Ещё это можно выразить так `TelegramBot  extends TelegramBot`. Если имена развернуть, то уже должно быть видно, что написано что-то не то.

На самом деле вы неправильно определили ответственность класса.

> расширяет функционал {@link TelegramBot}.

Покажите, пожалуйста, в каком конкретно месте он это делает. Я не вижу ни одного. И не надо в JavaDoc писать, что вы что-то "расширяете", слово extends и так все прекрасно видят. Пишите какую задачу выполняет класс, а не как он **реализован**.

> Он инициализируется с помощью токена, полученного из конфигурации приложения, и обрабатывает входящие обновления от Telegram, используя цепочку фильтров.

тоже пошли детали реализации. Их написать можно, ничего плохого в этом нет. Но **только после описания ответственности класса.**

Немного помогу: ответственность этого класса - настроить и запустить телеграмм бота. Ничего расширять не надо, просто создайте поле, настройте его и запустите. Назвать класс нужно соответствующе, то это будет совсем не "телеграмм бот".

LARCH1K replied:
Исправлено на взаимодействие с TelegramBot в поле

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/controller/TgBot.java:
```diff
@@ -0,0 +1,75 @@
+     * @param context контекст приложения Spring, используется для получения конфигурации и других бинов.
+     */
+    public TgBot(ApplicationContext context) {
+        super(context.getBean(ApplicationConfig.class).telegramToken());
```
мда.... неужели на курсе "Промышленная разработка" не учат определять и использовать бины?

LARCH1K replied:
Исправлено

vpyzhyanov replied:
В каком месте исправлено? Покажите, не вижу

![image](https://github.com/user-attachments/assets/4a628f52-1ab9-4099-b169-ae550ef6d292)

LARCH1K replied:
Исправили.  `ApplicationContext` оставляли для решения проблемы с циклической зависимостью. Сейчас решили проблему с использованием `@Lazy`

vpyzhyanov commented on pom.xml:
```diff
@@ -8,10 +8,52 @@
+
+            <dependency>
+                <groupId>com.github.pengrad</groupId>
+                <artifactId>java-telegram-bot-api</artifactId>
```
Интересно, почему выбор пал именно на эту библиотеку, а на ту, которую предлагает сам телеграмм? https://core.telegram.org/bots/tutorial

LARCH1K replied:
Данная библиотека была выбрана , так как уже был опыт работы с ней

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/controller/TgBot.java:
```diff
@@ -0,0 +1,75 @@
+                new Commands().getCommands()
+        ));
+
+        this.setUpdatesListener(updates -> {
```
обычно сначала идёт настройка, а потом запуск. А не наоборот

LARCH1K replied:
Настройка перенесена в конструктор

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/chain/BotFilterChain.java:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * Список фильтров, которые будут применяться последовательно.
+     */
+    private final List<ABotFilter> filters;
```
вот это тот самый момент, когда реально нужен интерфейс, абстрактного класса недостаточно. Если бы вы применяли ниспадающее программирование, то тут был бы интерфейс

LARCH1K replied:
В ходе исправления отошли от использования цепочки фильтров в пользу более простой обработки(относиться ко всем последующим замечаниям связанным с цепочкой фильтров)

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/chain/BotFilterChain.java:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * Класс BotFilterChain отвечает за создание и управление цепочкой фильтров для обработки обновлений Telegram.
+ * Цепочка фильтров выполняется последовательно, каждый фильтр передает управление следующему.
```
Реально шаблон "фильтр" применили? Очень интересно как это будет работать. И что ещё более интересно как вы это дальше будете расширять.

Я удивлён, ведь для такой задачи этот паттерн неприменим. Неужели вам удалось сломать стандарты и сделать невозможное?

vpyzhyanov replied:
А ещё это очень неэффективное решение

LARCH1K replied:
Отказались от использования фильтров

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/chain/BotFilterChain.java:
```diff
@@ -0,0 +1,55 @@
+     * Каждый фильтр будет вызывать следующий.
+     */
+    @PostConstruct
+    private void configureFilterChain() {
```
добро пожаловать на тёмную сторону. Коллективная безотстветственность. Задачу по инициализации класса вам удалось раскидать по разным местам.

PostConstruct уместен только тогда, когда вам реально нужно, чтобы весь спринг контекст был уже проинициализирован. В вашем случае - это не так. Вам можно всё сделать в конструкторе.

LARCH1K replied:
Класс удалён, в дальнейшем будем внимательнее с `@PostConstruct`

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/chain/BotFilterChain.java:
```diff
@@ -0,0 +1,55 @@
+    @PostConstruct
+    private void configureFilterChain() {
+        for (int i = 0; i < filters.size() - 1; i++) {
+            filters.get(i).setNextFilter(filters.get(i + 1));
```
очень интересно как вы определяете последовательность. ... Увидел, что через @Order

очень интересно как вы дальше будете расширять функционал, особенно интересно что будет, когда у вас появиться контекст. Например, одна и та же команда в разном контексте должна отрабатывать по разному или вообще не отрабатывать. Подозреваю, что вы будете решать кучу проблем, которые можно было не решать, а всё из-за того, что вы применили неправильный шаблон.

LARCH1K replied:
Отказались от использования фильтров

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/ABotFilter.java:
```diff
@@ -0,0 +1,62 @@
+import com.pengrad.telegrambot.model.Update;
+
+/**
+ * Абстрактный класс ABotFilter представляет собой элемент цепочки фильтров для обработки обновлений Telegram.
```
Не надо "воды" в описании класса. Все и так видят, что это абстрактный класс и все прекрасно видят как он называется

LARCH1K replied:
JavaDoc удалён вместе с классом

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/ABotFilter.java:
```diff
@@ -0,0 +1,62 @@
+package ru.naumen.bot.telegramBot.filter;
+
+import com.pengrad.telegrambot.model.Update;
```
не замечание, просто предупреждение. Надеюсь вы понимаете, что теперь с телеграмма вам не слезть никогда.

vpyzhyanov replied:
Видимо не понимаете, раз не слезли

hopenz replied:
Вроде бы слезли, сейчас телеграмм упоминается в двух классах (`TelegramBotController` и `ApplicationConfig`)

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/ABotFilter.java:
```diff
@@ -0,0 +1,62 @@
+
+/**
+ * Абстрактный класс ABotFilter представляет собой элемент цепочки фильтров для обработки обновлений Telegram.
+ * Каждый фильтр может передавать обработку следующему фильтру в цепочке, если он не завершил обработку сам.
```
Ну всё понятно, оказывается вы сломали сам шаблон "Фильтр", теперь это уже не фильтр, а что-то другое. То есть получается максимально запутали и себя и всю команду разработки

LARCH1K replied:
Отказались от использования фильтров

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/ABotFilter.java:
```diff
@@ -0,0 +1,62 @@
+ * Абстрактный класс ABotFilter представляет собой элемент цепочки фильтров для обработки обновлений Telegram.
+ * Каждый фильтр может передавать обработку следующему фильтру в цепочке, если он не завершил обработку сам.
+ */
+public abstract class ABotFilter {
```
Вам приходится платить за каждый символ в коде? К чему такая экономия букв? Или название должно быть TheBotFilter?))))

`AbstractBotFilter` или `BotFilterBase`

hopenz replied:
Данный класс удален, но название ABotFilter предполагало, что A - это предлог.
У нас с английским плохо видимо :)

vpyzhyanov replied:
Предлоги в названии классов не делают

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/ABotFilter.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public void doFilter(Update update) {
+        if (nextFilter != null) {
+            nextFilter.doFilter(update);
```
Вот вы и попались в ловушку недоделанной абстракции. Вроде писали абстрактный класс, но он получился вообще не абстрактным. Абстрактного мышления не хватает, надо прокачивать.

А теперь по существу, в чём конкретно проблема:

- Нет никакого контракта по которому должна работать реализация `ABotFilter`
- Нужно каким-то образом догадаться, что надо переопределить метод `doFilter`.  Каким - непонятно, даже в JavaDoc это не написано.
- Если забыть переопределить метод doFilter, то просто управление перейдёт следующему фильтру и ничего не будет сделано
- Если что-то можно забыть, то это **обязательно** будет забыто
- Даже если мы не забыли переопределить метод, то нужно не забыть вызвать `getNextFilter().doFilter(update)` или `super.doFilter(update)`
- Если что-то можно забыть, то это **обязательно** будет забыто (особенно если про это **НИГДЕ** не написано)

Попробуйте сами решить это проблему и написать настоящую абстракцию, которая решит все эти проблемы. Это вам будет очень полезно. Если что я подскажу как надо, но надо обязательно попробовать самим

vpyzhyanov replied:
пробовали?

hopenz replied:
Думали заменить абстрактный класс на интерфейс с методом doFilter, но так как решили отойти от использования фильтров применять не стали

vpyzhyanov replied:
очень зря, хорошая задачка, могли бы прокачаться в работе с абстракциями

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/ABotFilter.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return следующий фильтр.
+     */
+    public ABotFilter getNextFilter() {
```
прощай инкапсуляция. А самый прикол в том, что здесь незачем приносить её в жертву. Эта жертва совершенно напрасна.

LARCH1K replied:
Класс удалён

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/filterImpl/CommandFilter.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    /**
+     * Выполняет фильтрацию обновления.
```
Страшная ошибка, которая полностью обесценивает полиморфизм. Переопределяя описание метода, вы полностью обесцениваете исходный интерфейс.

LARCH1K replied:
JavaDoc удалён вместе с классом

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/filterImpl/MessageFilter.java:
```diff
@@ -0,0 +1,47 @@
+    public void doFilter(Update update) {
+        botProcessor.process(update);
+        if (this.getNextFilter() != null) {
+            this.getNextFilter().doFilter(update);
```
дублирование кода, не находите? Какая это тёмная сторона?

vpyzhyanov replied:
Нет ответа на вопрос!

LARCH1K replied:
Получается сторона "Слишком много абстракций". Хотели избежать дублирования использовав абстрактный класс, но сделали только хуже.

Класс удалён

vpyzhyanov replied:
Нет. Какая же тут абстракция? Тут конкретный класс.

Тут тёмная сторона - коллективная безответственность. Когда одно и то же делают разные классы. Отсюда и дублирование.

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/filterImpl/StartFilter.java:
```diff
@@ -0,0 +1,61 @@
+import static ru.naumen.bot.telegramBot.command.Commands.START_COMMAND;
+
+/**
+ * Класс StartFilter реализует фильтр для обработки команды "/start".
```
Интересно, чем же отличается команда start от всех остальных команд?

Ах да, добро пожаловать на тёмную сторону. Какую именно?

vpyzhyanov replied:
Нет ответа на вопрос!

hopenz replied:
Тут мы зашли на темную сторону "Коллективная безответственность". 
Обработка команды /start происходит так же, как и у остальных команд.

Данный класс удален.

vpyzhyanov replied:
Правильно

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/filter/filterImpl/StartFilter.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /**
+     * Выполняет фильтрацию обновления.
```
опять полиморфизм уничтожен

hopenz replied:
JavaDoc удалён вместе с классом

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/processor/BotProcessor.java:
```diff
@@ -0,0 +1,20 @@
+ * Интерфейс BotProcessor определяет контракт для обработки обновлений Telegram.
+ * Любой класс, реализующий этот интерфейс, должен предоставлять логику обработки входящих сообщений и событий.
+ */
+public interface BotProcessor {
```
В тему последнего занятия, скажите пожалуйста, где используется этот интерфейс?

Добро пожаловать на тёмную сторону ООП. На какую?

LARCH1K replied:
Интерфейс убран

vpyzhyanov replied:
Нет ответа на вопрос!

hopenz replied:
Данный интерфейс нигде не использовался.
Мы зашли на темную сторону "Слишком много абстракций" 

Класс удален

vpyzhyanov replied:
правильно

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/processor/BotProcessor.java:
```diff
@@ -0,0 +1,20 @@
+
+    /**
+     * Обрабатывает обновление от Telegram.
+     * Метод, который должен реализовать конкретную логику обработки обновлений.
```
спасибо КЭП)))

LARCH1K replied:
JavaDoc убран вместе с интерфейсом

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/service/BotService.java:
```diff
@@ -0,0 +1,41 @@
+ * Сервис BotService предоставляет основные методы для работы с Telegram ботом.
+ */
+@Service
+public class BotService {
```
Добро пожаловать в мир тёмных никому не нужных абстракций. Этот сервис просто посредник между возвращаемым сообщением и TgBot. По сути ничего ценного он не делает.

LARCH1K replied:
Класс удалён

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/service/BotService.java:
```diff
@@ -0,0 +1,41 @@
+     * @param message текст сообщения для отправки.
+     * @param update  обновление от Telegram, содержащее информацию о чате.
+     */
+    public void sendMessage(String message, Update update) {
```
А тут сервис делает то, что должен делать бот. Добро пожаловать в коллективную безотстветственность.

Отправить сообщение сам бот в состоянии и без посредников

LARCH1K replied:
Метод перенесён в TelegramBotController

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/processor/impl/MessageBotProcessor.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /**
+     * Добавляет доход и отправляет соответствующее собщение пользователю.
```
опечатка

LARCH1K replied:
Исправлено

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/processor/impl/MessageBotProcessor.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /**
+     * Добавляет расход и отправляет соответствующее собщение пользователю.
```
и тут

LARCH1K replied:
Исправлено

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/processor/impl/MessageBotProcessor.java:
```diff
@@ -0,0 +1,93 @@
+     * Добавляет расход и отправляет соответствующее собщение пользователю.
+     *
+     * @param expense сообщение пользователя.
+     * @param update  обноление от Telegram.
```
и походу везде

LARCH1K replied:
Исправлено

vpyzhyanov commented on bot/src/main/java/ru/naumen/bot/telegramBot/processor/impl/MessageBotProcessor.java:
```diff
@@ -0,0 +1,93 @@
+     */
+    private void addExpense(String expense, Update update) {
+        expenseService.addExpense(expense, update);
+        botService.sendMessage("Расход успешно добавлен!", update);
```
Похоже перегруз по ответственности вы тоже не обошли стороной. Никогда бы не подумал, что метод `addExpense` будет какие-то сообщения в чат отправлять.

Методы `addExpense` и `addIncome` вообще не нужны.

LARCH1K replied:
Методы убраны, логика перенесена в processMessage()

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/BotApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Test
+    void contextLoads() {
+    }
```
это чтобы тесты работали подольше? Поднимать весь спринговый контекст - ради чего?

LARCH1K replied:
Тест удалён

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/data/dao/inMemory/InMemoryBalanceDaoTest.java:
```diff
@@ -0,0 +1,26 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class InMemoryBalanceDaoTest {
```
код без JavaDoc приниматься не будет

LARCH1K replied:
JavaDoc ко всем тестам добавлен

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/data/dao/inMemory/InMemoryBalanceDaoTest.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @Test
+    void testSetBalanceAndGetBalance() {
```
код без JavaDoc приниматься не будет

LARCH1K replied:
JavaDoc ко всем тестам добавлен

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/telegramBot/command/CommandsTest.java:
```diff
@@ -0,0 +1,25 @@
+public class CommandsTest {
+
+    @Test
+    void testGetCommands() {
```
смысла в тесте ноль. Он просто дублирует реализацию

LARCH1K replied:
Тест удалён

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/telegramBot/filter/chain/BotFilterChainTest.java:
```diff
@@ -0,0 +1,47 @@
+
+        Method method = BotFilterChain.class.getDeclaredMethod("configureFilterChain");
+        method.setAccessible(true);
+        method.invoke(botFilterChain);
```
жесть..... не ну если стремитесь максимально усложнить всем жизнь, то норм. Но если нет, то тут вы должны были задуматься а всё ли правильно вы сделали.

LARCH1K replied:
Тест удалён

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/telegramBot/filter/chain/BotFilterChainTest.java:
```diff
@@ -0,0 +1,47 @@
+        method.invoke(botFilterChain);
+
+        doAnswer(invocation -> {
+            filter2.doFilter(invocation.getArgument(0));
```
тест вообще ничего не тестирует. Сами **в тесте** определили реализацию, сами её вызывали и проверили. Только смысла в этом нет

LARCH1K replied:
Тест удалён

vpyzhyanov commented on bot/src/test/java/ru/naumen/bot/telegramBot/filter/impl/CommandFilterTest.java:
```diff
@@ -0,0 +1,64 @@
+    @BeforeEach
+    void setUp() {
+        commandBotProcessor = mock(CommandBotProcessor.class);
+        update = mock(Update.class);
```
вот если бы не завязывались на телеграмм, то тесты были бы существенно проще. Ну ладно, сами выбрали путь без интерфейсов, пожинайте плоды

LARCH1K replied:
Тест удалён

vpyzhyanov replied:
но "плоды" остались

hopenz replied:
Максимально убрали зависимость классов от телеграмм.
Данный тест удален

vpyzhyanov reviewed (changes requested):
**Код без JavaDoc приниматься не будет!!!** В следующий раз даже смотреть не буду, сразу заверну!

Конечно вы полностью погрязли в тёмной стороне и нарушили всё, что только можно. Но в целом задача сделана довольно хорошо, видно, что старались сделать хорошо. Уверен в следующей задаче получится гораздо лучше.

LARCH1K committed:
Сделал проект одномодульным

LARCH1K committed:
Исправлен .gitignore

LARCH1K committed:
Исправлена структура приложения

hopenz committed:
Исправление JavaDocs

LARCH1K committed:
Исправлены тесты, добавлены javadoc для тестов

LARCH1K committed:
Merge branch 'task1-petrushina' into task1-ovchinnikov

LARCH1K committed:
Исправлены опечатки

hopenz committed:
Merge pull request #11 from hopenz/task1-ovchinnikov

Task1 ovchinnikov

LARCH1K commented:
Замечания исправлены.
Основные изменения: 
- отказались от обработки `update` в цепочке фильтров
- отказались от проброса `update` до уровня сервисов (теперь `update` доходит до `updateBotProcessor`, далее взаимодействие происходит с `String message, long chatId`)
- работа с `TelegramBot` перенесена в класс `TelegramBotController`

YaEtoTui commented:
1 задача принята. Я забыл написать( Чтобы было в истории, напишу, что я принял работу

LARCH1K committed:
Вынесли проверку, открыт ли чат с пользователем или будет открыт после обработки текущего сообщения, из UpdateBotProcessor в метод isChatActiveOrStarting класса CommandBotProcessor.

LARCH1K committed:
Merge branch 'task1-ovchinnikov' into task1

hopenz committed:
Исправление JavaDoc у BalanceDao и InMemoryUserDao

hopenz committed:
Merge pull request #12 from hopenz/task1-petrushina

Исправление JavaDoc у BalanceDao и InMemoryUserDao

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/command/Commands.java:
```diff
@@ -0,0 +1,72 @@
+package ru.naumen.bot.telegramBot.command;
```
Как команды связанные с телеграммом? Подозреваю что никак. Зачем вы тогда навязываете эту искуственную связь?

LARCH1K replied:
Команды перенесены в отдельный пакет

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/command/Commands.java:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @return массив команд, каждая из которых описывает действие и текстовое описание.
+     */
+    public BotCommand[] getCommands() {
```
а могли не мучаться, а сделать enum. Всё было бы в разы проще. Там даже поле с описанием можно добавить.

hopenz replied:
Теперь команды перечислены в enum, где содержится их текст команды и описание

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/processor/CommandBotProcessor.java:
```diff
@@ -0,0 +1,180 @@
+package ru.naumen.bot.telegramBot.processor;
```
Как все эти классы связанны с телеграммом? Подозреваю что никак. Зачем вы тогда навязываете эту искуственную связь?

LARCH1K replied:
Процессоры перенесены в отдельный пакет

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/service/BalanceService.java:
```diff
@@ -0,0 +1,34 @@
+package ru.naumen.bot.telegramBot.service;
```
Как все эти классы связанны с телеграммом? Подозреваю что никак. Зачем вы тогда навязываете эту искуственную связь?

LARCH1K replied:
Сервисы перенесены в отдельный пакет

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/controller/TelegramBotController.java:
```diff
@@ -0,0 +1,69 @@
+package ru.naumen.bot.telegramBot.controller;
```
В слове `telegramBot` слово Bot явно лишнее, т.к. оно уже в пути содержится.

Нужно стараться делать пакеты из **одного** слова. А если их несколько, то их разделять точкой.

hopenz replied:
Принято! название пакета изменено на "telegram"

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/controller/TelegramBotController.java:
```diff
@@ -0,0 +1,69 @@
+                new Commands().getCommands()
+        ));
+        this.telegramBot.setUpdatesListener(updates -> {
+            updates.forEach(applicationContext.getBean(UpdateBotProcessor.class)::processUpdate);
```
> мда.... неужели на курсе "Промышленная разработка" не учат определять и использовать бины?

Не исправлено!

LARCH1K replied:
Исправили

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/controller/TelegramBotController.java:
```diff
@@ -0,0 +1,69 @@
+     * и настраивает слушатель обновлений для обработки входящих обновлений.
+     *
+     * @param applicationConfig  конфигурация, содержащая токен бота Telegram
+     * @param applicationContext контекст приложения, используемый для получения бинов
```
убираем это

LARCH1K replied:
Убрали

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/controller/TelegramBotController.java:
```diff
@@ -0,0 +1,69 @@
+     * @param applicationConfig  конфигурация, содержащая токен бота Telegram
+     * @param applicationContext контекст приложения, используемый для получения бинов
+     */
+    public TelegramBotController(ApplicationConfig applicationConfig, ApplicationContext applicationContext) {
```
Убираем ApplicationContext

LARCH1K replied:
Убрали

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/processor/UpdateBotProcessor.java:
```diff
@@ -0,0 +1,54 @@
+package ru.naumen.bot.telegramBot.processor;
+
+import com.pengrad.telegrambot.model.Update;
```
домашнее задание на построение архитектуры чат-бота должно было научить вас не делать лишние связи. Очень жаль, что не научило.

Теперь из-за этого вы уже не слезете ни с телеграмма, ни даже с этой библиотеки (мало ли она поддерживаться перестанет)

LARCH1K replied:
Обработка `update` перенесена в `TelegramBotController`
Класс удалён

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/processor/UpdateBotProcessor.java:
```diff
@@ -0,0 +1,54 @@
+        long chatId = update.message().chat().id();
+        String message = update.message().text();
+        if (message != null) {
+            if (!commandBotProcessor.isChatActiveOrStarting(message, chatId)) return;
```
однострочное программирование - это очень дурной тон. Чтение и понимание кода это очень усложняет

hopenz replied:
Исправили

vpyzhyanov commented on src/main/java/ru/naumen/bot/telegramBot/processor/CommandBotProcessor.java:
```diff
@@ -0,0 +1,180 @@
+     */
+    public CommandBotProcessor(TelegramBotController telegramBotController, UserService userService, BalanceService balanceService,
+                               IncomeService incomeService, ExpenseService expenseService) {
+        this.telegramBotController = telegramBotController;
```
лишняя связь с телеграммом. Необходимость в ней точно отсутствует.

Хотя в данном случае только слово Telegramm меня смущает. Больше связей с телеграммом нет.

LARCH1K replied:
`TelegramBotController` теперь имплементирует интерфейс `BotController` . В процессорах теперь взаимодействие происходит с помощью интерфейса `BotController`

vpyzhyanov commented on src/test/java/ru/naumen/bot/telegramBot/processor/UpdateBotProcessorTest.java:
```diff
@@ -0,0 +1,147 @@
+     * @param messageText текст сообщения
+     * @return объект {@link Update} с заданными параметрами
+     */
+    private Update createUpdate(long chatId, String messageText) {
```
вот делать вам нечего. Сами усложнили себе жизнь. А могли без этих сложностей спокойно обойтись! Весь этот метод вообще не нужен

hopenz replied:
Метод удален вместе с классом

vpyzhyanov reviewed (changes requested):
Пока вы не начнёте отвечать на вопросы и исправлять замечания (с этим дела получше), то у вас не получится сдать задачу

LARCH1K committed:
Fix mistakes

* Команды изменены на enum
* Логика не связанная с телеграммом вынесена в отдельные пакеты
* Обработка update перенесена в TelegramBotController
* TelegramBotController имплементирует BotController

LARCH1K committed:
Merge pull request #13 from hopenz/task1-ovchinnikov

Fix mistakes

hopenz committed:
fix mistakes

* Поправила тесты
* Написала JavaDoc, где случайно пропустили
* Исправила косяки с public полями

hopenz committed:
Merge pull request #14 from hopenz/task1-petrushina

fix mistakes

LARCH1K commented:
Замечания исправили, на вопросы ответили

Основные изменения:
* Команды изменены на enum
* Логика не связанная с телеграммом вынесена в отдельные пакеты
* Обработка update перенесена в TelegramBotController
* TelegramBotController имплементирует BotController
* Процессоры используют BotController

vpyzhyanov commented on src/main/java/ru/naumen/bot/processor/CommandBotProcessor.java:
```diff
@@ -74,13 +72,28 @@ public CommandBotProcessor(TelegramBotController telegramBotController, UserServ
      */
     public void processCommand(String message, long chatId) {
-        switch (message) {
+        Arrays.stream(Commands.values())
```
не надо каждый раз перебирать все команды. Просто один раз при создании положите их в подходящую для этого коллекцию

LARCH1K replied:
Исправили, теперь при запуске приложения команды записываются в Map и в методе достаются по ключу, где ключ строковый текст команды

vpyzhyanov reviewed (approved):
Процесс обучения прошёл успешно. Задача принята. Можете мёрджить

vpyzhyanov commented:
Оценка 

- ООП: 5 из 5
- Git 3 из 3

hopenz committed:
Исправление замечания

hopenz committed:
Merge pull request #15 from hopenz/task1-petrushina

Исправление замечания
